File: sdks/php/src/MarvinPayClient.php
<?php

declare(strict_types=1);

namespace MarvinPay;

/**
 * Zero-dependency HTTP client for the Marvin Pay payment gateway.
 *
 * Transport is plain cURL + JSON. Authentication is the merchant `X-API-KEY`
 * header for `/v1/payment/**`; an optional portal JWT (`Authorization: Bearer`)
 * can be supplied for the JWT-gated surfaces.
 *
 * All field names on request bodies match the API contract's snake_case exactly.
 * Amounts are whole numbers (XAF/XOF have no minor units); the client serializes
 * the `amount` field without a fractional part.
 *
 * @see ../../CONTRACT.md  the authoritative API contract
 */
class MarvinPayClient
{
    public const DEFAULT_BASE_URL = 'https://api.marvincorporate.co/api';

    private string $apiKey;
    private string $baseUrl;
    private ?string $bearerToken;
    private int $timeout;

    /** @var array<string,string> lowercased header name => value from the last response */
    private array $lastResponseHeaders = [];

    /**
     * @param string               $apiKey  merchant API key sent as `X-API-KEY` (may be '' for public hosted-pay only)
     * @param array<string,mixed>  $options {
     *     base_url?:     string  default https://api.marvincorporate.co/api
     *     bearer_token?: string  optional portal JWT for JWT-gated endpoints
     *     timeout?:      int     request timeout in seconds (default 30)
     * }
     */
    public function __construct(string $apiKey, array $options = [])
    {
        $this->apiKey = $apiKey;
        $this->baseUrl = \rtrim((string) ($options['base_url'] ?? self::DEFAULT_BASE_URL), '/');
        $this->bearerToken = isset($options['bearer_token']) && $options['bearer_token'] !== ''
            ? (string) $options['bearer_token']
            : null;
        $this->timeout = (int) ($options['timeout'] ?? 30);
    }

    // ─────────────────────────────────────────────────────────────────────
    //  Core payment endpoints (X-API-KEY)
    // ─────────────────────────────────────────────────────────────────────

    /**
     * Collect from a customer (payer -> merchant). `POST /v1/payment/collect`.
     *
     * Always sends `X-Idempotency-Key`; when $idempotencyKey is null it defaults
     * to the body's `transaction_id`.
     *
     * @param array<string,mixed> $paymentRequest snake_case PaymentRequest
     * @return array<string,mixed> PaymentResult
     */
    public function collect(array $paymentRequest, ?string $idempotencyKey = null): array
    {
        return $this->moneyMove('/v1/payment/collect', $paymentRequest, $idempotencyKey);
    }

    /**
     * Pay out to a recipient (merchant -> recipient). `POST /v1/payment/payout`.
     *
     * @param array<string,mixed> $paymentRequest snake_case PaymentRequest
     * @return array<string,mixed> PaymentResult
     */
    public function payout(array $paymentRequest, ?string $idempotencyKey = null): array
    {
        return $this->moneyMove('/v1/payment/payout', $paymentRequest, $idempotencyKey);
    }

    /**
     * Authoritative status check. `GET /v1/payment/status/{transactionId}`.
     *
     * @return array<string,mixed> TransactionStatusResponse
     */
    public function getStatus(string $transactionId): array
    {
        return $this->request('GET', '/v1/payment/status/' . \rawurlencode($transactionId));
    }

    /**
     * Fee estimate. `GET /v1/payment/fees`.
     *
     * @param array<string,mixed> $params query params: currency, amount, direction, fee_bearer
     * @return array<string,mixed> FeeEstimateResponse
     */
    public function getFees(array $params): array
    {
        if (isset($params['amount']) && \is_numeric($params['amount'])) {
            $params['amount'] = (int) $params['amount'];
        }
        return $this->request('GET', '/v1/payment/fees', ['query' => $params]);
    }

    /**
     * List provider names valid for a country.
     * `GET /v1/payment/payment-methods/{countryCode}`.
     *
     * @return array<int,string> provider-name strings (the values you pass as payment_method)
     */
    public function getPaymentMethods(string $countryCode): array
    {
        return $this->request('GET', '/v1/payment/payment-methods/' . \rawurlencode($countryCode));
    }

    /**
     * Poll {@see getStatus()} until the transaction reaches a terminal state.
     *
     * Schedule (per the contract): first poll after 5s, then exponential backoff
     * capped at 60s, giving up after 600s (10 min). Returns the last
     * TransactionStatusResponse — inspect `transaction_status`
     * (`SUCCESSFUL` / `FAILED` / still `PENDING` on timeout).
     *
     * @param array<string,mixed> $opts {
     *     initial_delay?: int   first-poll delay in seconds (default 5)
     *     max_delay?:     int   backoff cap in seconds (default 60)
     *     timeout?:       int   overall budget in seconds (default 600)
     *     sleep?:         callable(int):void  injectable sleeper (default \sleep) for testing
     * }
     * @return array<string,mixed> the final TransactionStatusResponse
     */
    public function waitForCompletion(string $transactionId, array $opts = []): array
    {
        $delay = (int) ($opts['initial_delay'] ?? 5);
        $cap = (int) ($opts['max_delay'] ?? 60);
        $timeout = (int) ($opts['timeout'] ?? 600);
        /** @var callable(int):void $sleep */
        $sleep = $opts['sleep'] ?? static function (int $seconds): void {
            if ($seconds > 0) {
                \sleep($seconds);
            }
        };

        $deadline = \time() + $timeout;
        $last = null;

        while (true) {
            $remaining = $deadline - \time();
            if ($remaining <= 0) {
                break;
            }
            $sleep(\min($delay, $remaining));

            $last = $this->getStatus($transactionId);
            $status = \strtoupper((string) ($last['transaction_status'] ?? ''));
            if ($status === 'SUCCESSFUL' || $status === 'FAILED') {
                return $last;
            }

            $delay = \min($delay * 2, $cap);
        }

        // Timed out still pending — return the last observation (or fetch once).
        return $last ?? $this->getStatus($transactionId);
    }

    // ─────────────────────────────────────────────────────────────────────
    //  Public "hosted pay" flows (no API key required by the server)
    // ─────────────────────────────────────────────────────────────────────

    /**
     * Pay a public invoice. `POST /v1/invoices/{reference}/pay`.
     *
     * @param array<string,mixed> $payInvoiceRequest PayInvoiceRequest:
     *        country_code, currency, mobile_number, payment_method,
     *        beneficiary_name (required), customer_email (optional)
     * @return array<string,mixed> PaymentResult
     */
    public function payInvoice(string $reference, array $payInvoiceRequest): array
    {
        return $this->request('POST', '/v1/invoices/' . \rawurlencode($reference) . '/pay', [
            'json' => $this->normalizeAmount($payInvoiceRequest),
        ]);
    }

    /**
     * Contribute to a public campaign. `POST /v1/campaigns/{reference}/contribute`.
     *
     * @param array<string,mixed> $contributeRequest ContributeRequest:
     *        country_code, currency, amount (min 100), mobile_number,
     *        payment_method, beneficiary_name, customer_email
     * @return array<string,mixed> PaymentResult
     */
    public function contributeCampaign(string $reference, array $contributeRequest): array
    {
        return $this->request('POST', '/v1/campaigns/' . \rawurlencode($reference) . '/contribute', [
            'json' => $this->normalizeAmount($contributeRequest),
        ]);
    }

    /**
     * Pay a QR code. `POST /v1/merchant/qrcode/pay/{qrReference}`.
     * The merchant API key is resolved server-side from the QR reference.
     *
     * @param array<string,mixed> $qrPaymentRequest QRPaymentRequest:
     *        country_code, currency, amount (ignored if the QR has a fixedAmount),
     *        mobile_number, payment_method, beneficiary_name (required), customer_email
     * @return array<string,mixed> PaymentResult
     */
    public function payQr(string $qrReference, array $qrPaymentRequest): array
    {
        return $this->request('POST', '/v1/merchant/qrcode/pay/' . \rawurlencode($qrReference), [
            'json' => $this->normalizeAmount($qrPaymentRequest),
        ]);
    }

    /**
     * Public poll for a hosted-pay (QR/invoice/campaign) transaction.
     * `GET /v1/merchant/qrcode/status/{transactionId}`. Poll ~every 5s.
     *
     * @return array<string,mixed> plain map: transactionId, status, amount,
     *         currency, paymentMethod, mobileNumber, timestamp
     */
    public function getQrStatus(string $transactionId): array
    {
        return $this->request('GET', '/v1/merchant/qrcode/status/' . \rawurlencode($transactionId));
    }

    // ─────────────────────────────────────────────────────────────────────
    //  Response-header access (idempotency replay surfacing)
    // ─────────────────────────────────────────────────────────────────────

    /**
     * Headers from the most recent response, keyed by lowercased header name.
     * Use to read `x-idempotency-replay` / `x-idempotency-key-auto` after a
     * collect/payout.
     *
     * @return array<string,string>
     */
    public function getLastResponseHeaders(): array
    {
        return $this->lastResponseHeaders;
    }

    // ─────────────────────────────────────────────────────────────────────
    //  Internals
    // ─────────────────────────────────────────────────────────────────────

    /**
     * @param array<string,mixed> $body
     * @return array<string,mixed>
     */
    private function moneyMove(string $path, array $body, ?string $idempotencyKey): array
    {
        $body = $this->normalizeAmount($body);

        // Always send X-Idempotency-Key; default to transaction_id.
        $key = $idempotencyKey ?? ($body['transaction_id'] ?? null);
        $headers = [];
        if ($key !== null && $key !== '') {
            $headers['X-Idempotency-Key'] = (string) $key;
        }

        return $this->request('POST', $path, ['json' => $body, 'headers' => $headers]);
    }

    /**
     * Ensure `amount` serializes as a whole number (no fractional part).
     *
     * @param array<string,mixed> $body
     * @return array<string,mixed>
     */
    private function normalizeAmount(array $body): array
    {
        if (isset($body['amount']) && \is_numeric($body['amount'])) {
            $body['amount'] = (int) $body['amount'];
        }
        return $body;
    }

    /**
     * Perform an HTTP request with cURL. Retries once on 429/5xx for GETs only.
     *
     * @param array<string,mixed> $opts { json?: array, query?: array, headers?: array<string,string> }
     * @return array<mixed> decoded JSON body ([] for an empty/non-array 2xx body)
     */
    private function request(string $method, string $path, array $opts = []): array
    {
        $method = \strtoupper($method);
        $isGet = $method === 'GET';
        $maxAttempts = $isGet ? 2 : 1; // retry once on GET

        $attempt = 0;

        while (true) {
            $attempt++;
            [$status, $headers, $raw, $curlError] = $this->doRequest($method, $path, $opts);
            $this->lastResponseHeaders = $headers;

            if ($curlError !== null) {
                if ($isGet && $attempt < $maxAttempts) {
                    $this->backoff(1, 0);
                    continue;
                }
                throw new MarvinPayException('Marvin Pay request failed: ' . $curlError);
            }

            $retryable = $status === 429 || $status >= 500;
            if ($retryable && $isGet && $attempt < $maxAttempts) {
                $retryAfter = isset($headers['retry-after']) ? (int) $headers['retry-after'] : 0;
                $this->backoff($attempt, $retryAfter);
                continue;
            }

            $decoded = $this->decode($raw);

            if ($status < 200 || $status >= 300) {
                $message = $this->extractMessage($decoded, $raw, $status);
                throw new MarvinPayException($message, $status, $decoded ?? ($raw !== '' ? $raw : null));
            }

            // These endpoints return JSON objects/arrays; coerce an empty or
            // non-array 2xx body to [] so typed callers never see null.
            return \is_array($decoded) ? $decoded : [];
        }
    }

    /**
     * Single cURL round-trip.
     *
     * @param array<string,mixed> $opts
     * @return array{0:int,1:array<string,string>,2:string,3:?string} [status, headers, rawBody, curlError]
     */
    private function doRequest(string $method, string $path, array $opts): array
    {
        $url = $this->baseUrl . $path;
        if (!empty($opts['query'])) {
            $qs = \http_build_query($opts['query']);
            if ($qs !== '') {
                $url .= (\str_contains($url, '?') ? '&' : '?') . $qs;
            }
        }

        $headers = ['Accept: application/json'];
        if ($this->apiKey !== '') {
            $headers[] = 'X-API-KEY: ' . $this->apiKey;
        }
        if ($this->bearerToken !== null) {
            $headers[] = 'Authorization: Bearer ' . $this->bearerToken;
        }
        foreach (($opts['headers'] ?? []) as $name => $value) {
            $headers[] = $name . ': ' . $value;
        }

        $bodyJson = null;
        if (\array_key_exists('json', $opts)) {
            $bodyJson = \json_encode($opts['json'], JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE);
            $headers[] = 'Content-Type: application/json';
        }

        $ch = \curl_init();
        $responseHeaders = [];

        \curl_setopt_array($ch, [
            CURLOPT_URL => $url,
            CURLOPT_CUSTOMREQUEST => $method,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_HTTPHEADER => $headers,
            CURLOPT_TIMEOUT => $this->timeout,
            CURLOPT_CONNECTTIMEOUT => \min($this->timeout, 15),
            CURLOPT_HEADERFUNCTION => function ($ch, string $header) use (&$responseHeaders): int {
                $len = \strlen($header);
                $parts = \explode(':', $header, 2);
                if (\count($parts) === 2) {
                    $responseHeaders[\strtolower(\trim($parts[0]))] = \trim($parts[1]);
                }
                return $len;
            },
        ]);

        if ($bodyJson !== null) {
            \curl_setopt($ch, CURLOPT_POSTFIELDS, $bodyJson);
        }

        $raw = \curl_exec($ch);
        $errno = \curl_errno($ch);
        $error = $errno !== 0 ? \curl_error($ch) : null;
        $status = (int) \curl_getinfo($ch, CURLINFO_RESPONSE_CODE);
        \curl_close($ch);

        return [$status, $responseHeaders, \is_string($raw) ? $raw : '', $error];
    }

    /**
     * @return mixed decoded array/scalar, or null for empty/undecodable bodies
     */
    private function decode(string $raw)
    {
        if ($raw === '') {
            return null;
        }
        $decoded = \json_decode($raw, true);
        return \json_last_error() === JSON_ERROR_NONE ? $decoded : null;
    }

    /**
     * @param mixed $decoded
     */
    private function extractMessage($decoded, string $raw, int $status): string
    {
        if (\is_array($decoded)) {
            if (isset($decoded['message']) && \is_string($decoded['message']) && $decoded['message'] !== '') {
                return $decoded['message'];
            }
            if (isset($decoded['error']) && \is_string($decoded['error']) && $decoded['error'] !== '') {
                return $decoded['error'];
            }
        }
        if ($raw !== '') {
            return 'HTTP ' . $status . ': ' . $raw;
        }
        return 'HTTP ' . $status;
    }

    private function backoff(int $attempt, int $retryAfterSeconds): void
    {
        $seconds = $retryAfterSeconds > 0 ? $retryAfterSeconds : \min(2 ** $attempt, 5);
        \sleep($seconds);
    }
}

File: sdks/java/src/main/java/co/marvincorporate/pay/sdk/MarvinPayClient.java
package co.marvincorporate.pay.sdk;

import co.marvincorporate.pay.sdk.model.Direction;
import co.marvincorporate.pay.sdk.model.FeeBearer;
import co.marvincorporate.pay.sdk.model.FeeEstimate;
import co.marvincorporate.pay.sdk.model.PaymentRequest;
import co.marvincorporate.pay.sdk.model.PaymentResult;
import co.marvincorporate.pay.sdk.model.TransactionStatusResponse;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.IOException;
import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Thin, self-contained client for the Marvin Pay payment gateway.
 *
 * <p>Built on the JDK {@link HttpClient} and Jackson only. Thread-safe and cheap to
 * reuse — create one instance per API key and share it.
 *
 * <h2>Authentication</h2>
 * <ul>
 *   <li>{@code X-API-KEY} is sent on the authenticated payment surface when an
 *       {@code apiKey} is configured (collect, payout, status, fees, payment-methods).</li>
 *   <li>{@code Authorization: Bearer} is sent when a {@code bearerToken} is configured
 *       (only needed for JWT/portal endpoints, which this thin client does not expose
 *       beyond passing the header).</li>
 *   <li>The hosted-pay helpers ({@code payInvoice}, {@code contributeCampaign},
 *       {@code payQr}, {@code getQrStatus}) are <b>public</b> and send no auth header.</li>
 * </ul>
 *
 * <h2>Idempotency</h2>
 * Every {@code collect}/{@code payout} sends {@code X-Idempotency-Key}. If you do not
 * pass one explicitly it defaults to the request's {@code idempotency_key}, and failing
 * that to its {@code transaction_id}.
 *
 * <h2>Errors &amp; retries</h2>
 * Non-2xx responses raise {@link MarvinPayException} (with HTTP status + body). GETs are
 * retried once on {@code 429}/{@code 5xx} (respecting {@code Retry-After} when present);
 * money-moving POSTs are never auto-retried.
 */
public class MarvinPayClient {

    private final MarvinPayConfig config;
    private final HttpClient http;
    private final ObjectMapper mapper;

    /** Convenience constructor: production base URL, API-key auth, 30s timeout. */
    public MarvinPayClient(String apiKey) {
        this(MarvinPayConfig.builder().apiKey(apiKey).build());
    }

    public MarvinPayClient(MarvinPayConfig config) {
        if (config == null) {
            throw new IllegalArgumentException("config must not be null");
        }
        this.config = config;
        this.http = HttpClient.newBuilder()
                .connectTimeout(config.getTimeout())
                .build();
        this.mapper = new ObjectMapper()
                .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
    }

    // =====================================================================
    // Core payment surface (X-API-KEY)
    // =====================================================================

    /** {@code POST /v1/payment/collect} — collect from a customer (payer → merchant). */
    public PaymentResult collect(PaymentRequest req) {
        return collect(req, null);
    }

    /** As {@link #collect(PaymentRequest)} but with an explicit idempotency key. */
    public PaymentResult collect(PaymentRequest req, String idempotencyKey) {
        return postPayment("/v1/payment/collect", req, idempotencyKey);
    }

    /** {@code POST /v1/payment/payout} — pay out (merchant → recipient). */
    public PaymentResult payout(PaymentRequest req) {
        return payout(req, null);
    }

    /** As {@link #payout(PaymentRequest)} but with an explicit idempotency key. */
    public PaymentResult payout(PaymentRequest req, String idempotencyKey) {
        return postPayment("/v1/payment/payout", req, idempotencyKey);
    }

    /** {@code GET /v1/payment/status/{transactionId}} — the authoritative outcome check. */
    public TransactionStatusResponse getStatus(String transactionId) {
        requireNonBlank(transactionId, "transactionId");
        return get("/v1/payment/status/" + encodePathSegment(transactionId),
                true, TransactionStatusResponse.class);
    }

    /**
     * {@code GET /v1/payment/fees} — fee estimate / fee-bearer split.
     *
     * @param direction {@code COLLECT} or {@code PAYOUT} (also {@code TOPUP}/{@code WITHDRAWAL}).
     * @param feeBearer {@code MERCHANT}/{@code CUSTOMER}, or {@code null} to omit (server default MERCHANT).
     */
    public FeeEstimate getFees(String currency, long amount, String direction, String feeBearer) {
        requireNonBlank(currency, "currency");
        requireNonBlank(direction, "direction");
        StringBuilder path = new StringBuilder("/v1/payment/fees")
                .append("?currency=").append(encodeQuery(currency))
                .append("&amount=").append(amount)
                .append("&direction=").append(encodeQuery(direction));
        if (feeBearer != null && !feeBearer.isBlank()) {
            path.append("&fee_bearer=").append(encodeQuery(feeBearer));
        }
        return get(path.toString(), true, FeeEstimate.class);
    }

    /** Type-safe overload of {@link #getFees(String, long, String, String)}. */
    public FeeEstimate getFees(String currency, long amount, Direction direction, FeeBearer feeBearer) {
        return getFees(currency, amount,
                direction == null ? null : direction.name(),
                feeBearer == null ? null : feeBearer.name());
    }

    /**
     * {@code GET /v1/payment/payment-methods/{countryCode}} — provider-name strings valid
     * for the country (the values to pass as {@code payment_method}, e.g. {@code mtn_cm}).
     */
    public List<String> getPaymentMethods(String countryCode) {
        requireNonBlank(countryCode, "countryCode");
        String[] methods = get("/v1/payment/payment-methods/" + encodePathSegment(countryCode),
                true, String[].class);
        return methods == null ? List.of() : Arrays.asList(methods);
    }

    // =====================================================================
    // Waiting for completion (poll GET /v1/payment/status)
    // =====================================================================

    /**
     * Poll {@code getStatus} until the transaction is terminal or the timeout elapses,
     * using the SDK defaults: first poll after <b>5s</b>, exponential backoff capped at
     * <b>60s</b>, give up after <b>10 minutes</b>.
     *
     * @return the last status observed. If it timed out this is still {@code PENDING};
     *         check {@link TransactionStatusResponse#isTerminal()}.
     */
    public TransactionStatusResponse waitForCompletion(String transactionId) {
        return waitForCompletion(transactionId,
                Duration.ofSeconds(5), Duration.ofSeconds(60), Duration.ofMinutes(10));
    }

    /**
     * Poll {@code getStatus} until the transaction reaches a terminal state
     * ({@code SUCCESSFUL}/{@code FAILED}) or {@code timeout} elapses.
     *
     * <p>Sleeps {@code startDelay} before the first poll, then doubles the delay after
     * each still-pending poll, capped at {@code maxDelay}.
     *
     * @return the last status observed (still {@code PENDING} if it timed out).
     * @throws MarvinPayException if the thread is interrupted while waiting.
     */
    public TransactionStatusResponse waitForCompletion(String transactionId,
                                                       Duration startDelay,
                                                       Duration maxDelay,
                                                       Duration timeout) {
        requireNonBlank(transactionId, "transactionId");
        long deadlineNanos = System.nanoTime() + timeout.toNanos();
        Duration delay = startDelay;
        TransactionStatusResponse last = null;
        while (true) {
            sleep(delay);
            last = getStatus(transactionId);
            if (last.isTerminal()) {
                return last;
            }
            if (System.nanoTime() >= deadlineNanos) {
                return last; // timed out; still pending
            }
            long nextMillis = Math.min(delay.toMillis() * 2, maxDelay.toMillis());
            delay = Duration.ofMillis(Math.max(nextMillis, 1));
        }
    }

    // =====================================================================
    // Public "hosted pay" helpers (no API key sent; auth resolved server-side)
    // =====================================================================

    /**
     * {@code POST /v1/invoices/{reference}/pay}. Populate {@link PaymentRequest} with
     * {@code countryCode}, {@code currency}, {@code mobileNumber}, {@code paymentMethod},
     * {@code beneficiaryName} (required), and optionally {@code customerEmail}. The invoice
     * already carries the amount, so leave {@code amount} unset.
     */
    public PaymentResult payInvoice(String reference, PaymentRequest req) {
        requireNonBlank(reference, "reference");
        return post("/v1/invoices/" + encodePathSegment(reference) + "/pay",
                req, null, false, PaymentResult.class);
    }

    /**
     * {@code POST /v1/campaigns/{reference}/contribute}. Set {@code amount} (min 100) plus
     * the payer fields ({@code countryCode}, {@code currency}, {@code mobileNumber},
     * {@code paymentMethod}, {@code beneficiaryName}, optional {@code customerEmail}).
     */
    public PaymentResult contributeCampaign(String reference, PaymentRequest req) {
        requireNonBlank(reference, "reference");
        return post("/v1/campaigns/" + encodePathSegment(reference) + "/contribute",
                req, null, false, PaymentResult.class);
    }

    /**
     * {@code POST /v1/merchant/qrcode/pay/{qrReference}}. {@code amount} is ignored if the
     * QR has a fixed amount. The merchant API key is resolved server-side from the QR
     * reference — never send it.
     */
    public PaymentResult payQr(String qrReference, PaymentRequest req) {
        requireNonBlank(qrReference, "qrReference");
        return post("/v1/merchant/qrcode/pay/" + encodePathSegment(qrReference),
                req, null, false, PaymentResult.class);
    }

    /**
     * {@code GET /v1/merchant/qrcode/status/{transactionId}} — public poll used by the
     * QR/invoice/campaign pay pages. Returns a plain map with keys {@code transactionId},
     * {@code status}, {@code amount}, {@code currency}, {@code paymentMethod},
     * {@code mobileNumber}, {@code timestamp}. No merchant/fee data is exposed. Poll ~every 5s.
     */
    public Map<String, Object> getQrStatus(String transactionId) {
        requireNonBlank(transactionId, "transactionId");
        return getMap("/v1/merchant/qrcode/status/" + encodePathSegment(transactionId), false);
    }

    // =====================================================================
    // Internals
    // =====================================================================

    private PaymentResult postPayment(String path, PaymentRequest req, String idempotencyKey) {
        if (req == null) {
            throw new IllegalArgumentException("PaymentRequest must not be null");
        }
        String key = firstNonBlank(idempotencyKey, req.getIdempotencyKey(), req.getTransactionId());
        Map<String, String> headers = new HashMap<>();
        if (key != null) {
            headers.put("X-Idempotency-Key", key);
        }
        return post(path, req, headers, true, PaymentResult.class);
    }

    private <T> T post(String path, Object body, Map<String, String> extraHeaders,
                       boolean auth, Class<T> responseType) {
        HttpRequest.Builder b = requestBuilder(path, auth)
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofByteArray(writeJson(body)));
        if (extraHeaders != null) {
            extraHeaders.forEach(b::header);
        }
        HttpResponse<String> resp = send(b.build(), false); // POSTs are not auto-retried
        return handle(resp, responseType);
    }

    private <T> T get(String path, boolean auth, Class<T> responseType) {
        HttpRequest req = requestBuilder(path, auth).GET().build();
        HttpResponse<String> resp = send(req, true); // GETs retry once on 429/5xx
        return handle(resp, responseType);
    }

    private Map<String, Object> getMap(String path, boolean auth) {
        HttpRequest req = requestBuilder(path, auth).GET().build();
        HttpResponse<String> resp = send(req, true);
        int code = resp.statusCode();
        if (code < 200 || code >= 300) {
            throw new MarvinPayException(code, extractMessage(resp.body(), code), resp.body());
        }
        String body = resp.body();
        if (body == null || body.isBlank()) {
            return Map.of();
        }
        try {
            return mapper.readValue(body, new TypeReference<Map<String, Object>>() {});
        } catch (IOException e) {
            throw new MarvinPayException("Failed to parse response JSON: " + e.getMessage(), e);
        }
    }

    private HttpRequest.Builder requestBuilder(String path, boolean auth) {
        HttpRequest.Builder b = HttpRequest.newBuilder(URI.create(config.getBaseUrl() + path))
                .timeout(config.getTimeout())
                .header("Accept", "application/json");
        if (auth) {
            if (config.getApiKey() != null) {
                b.header("X-API-KEY", config.getApiKey());
            }
            if (config.getBearerToken() != null) {
                b.header("Authorization", "Bearer " + config.getBearerToken());
            }
        }
        return b;
    }

    private HttpResponse<String> send(HttpRequest request, boolean retryable) {
        try {
            HttpResponse<String> resp = http.send(request,
                    HttpResponse.BodyHandlers.ofString(StandardCharsets.UTF_8));
            if (retryable && isRetryable(resp.statusCode())) {
                sleep(retryDelay(resp));
                resp = http.send(request, HttpResponse.BodyHandlers.ofString(StandardCharsets.UTF_8));
            }
            return resp;
        } catch (IOException e) {
            throw new MarvinPayException("HTTP request failed: " + e.getMessage(), e);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new MarvinPayException("HTTP request interrupted", e);
        }
    }

    private static boolean isRetryable(int status) {
        return status == 429 || status >= 500;
    }

    /** Honour {@code Retry-After} (seconds) when present, capped; else a small fixed backoff. */
    private static Duration retryDelay(HttpResponse<String> resp) {
        return resp.headers().firstValue("Retry-After")
                .map(String::trim)
                .flatMap(MarvinPayClient::parseLong)
                .map(secs -> Duration.ofSeconds(Math.min(Math.max(secs, 0), 30)))
                .orElse(Duration.ofSeconds(1));
    }

    private <T> T handle(HttpResponse<String> resp, Class<T> responseType) {
        int code = resp.statusCode();
        String body = resp.body();
        if (code < 200 || code >= 300) {
            throw new MarvinPayException(code, extractMessage(body, code), body);
        }
        if (responseType == null || responseType == Void.class) {
            return null;
        }
        return readJson(body, responseType);
    }

    /** Best-effort extraction of a human message from an error body. */
    private String extractMessage(String body, int code) {
        if (body != null && !body.isBlank()) {
            try {
                Map<String, Object> map = mapper.readValue(body, new TypeReference<Map<String, Object>>() {});
                Object msg = map.get("message");
                if (msg == null) {
                    msg = map.get("error");
                }
                if (msg != null && !msg.toString().isBlank()) {
                    return "HTTP " + code + ": " + msg;
                }
            } catch (IOException ignored) {
                // Non-JSON body; fall through.
            }
        }
        return "HTTP " + code;
    }

    private byte[] writeJson(Object body) {
        try {
            return mapper.writeValueAsBytes(body);
        } catch (IOException e) {
            throw new MarvinPayException("Failed to serialize request JSON: " + e.getMessage(), e);
        }
    }

    private <T> T readJson(String body, Class<T> type) {
        if (body == null || body.isBlank()) {
            throw new MarvinPayException("Expected a JSON body but the response was empty");
        }
        try {
            return mapper.readValue(body, type);
        } catch (IOException e) {
            throw new MarvinPayException("Failed to parse response JSON: " + e.getMessage(), e);
        }
    }

    private static void sleep(Duration d) {
        long millis = d == null ? 0 : Math.max(d.toMillis(), 0);
        if (millis == 0) {
            return;
        }
        try {
            Thread.sleep(millis);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new MarvinPayException("Interrupted while waiting", e);
        }
    }

    private static String encodePathSegment(String value) {
        // URLEncoder targets query strings (space → '+'); fix that for path segments.
        return URLEncoder.encode(value, StandardCharsets.UTF_8).replace("+", "%20");
    }

    private static String encodeQuery(String value) {
        return URLEncoder.encode(value, StandardCharsets.UTF_8);
    }

    private static String firstNonBlank(String... values) {
        if (values != null) {
            for (String v : values) {
                if (v != null && !v.isBlank()) {
                    return v;
                }
            }
        }
        return null;
    }

    private static void requireNonBlank(String value, String name) {
        if (value == null || value.isBlank()) {
            throw new IllegalArgumentException(name + " must not be null or blank");
        }
    }

    private static java.util.Optional<Long> parseLong(String s) {
        try {
            return java.util.Optional.of(Long.parseLong(s));
        } catch (NumberFormatException e) {
            return java.util.Optional.empty();
        }
    }
}
